ENIGMAsystem/SHELL/Platforms/SDL/Gamepad.cpp

```diff
@@ -0,0 +1,155 @@
+#include "Gamepad.h"
+
+#include <SDL2/SDL.h>
+
+#include <string>
+
+namespace enigma {
+  std::vector<GamepadState> gamepads;
+}
+
+namespace enigma_user {
+
+bool gamepad_is_supported() {
+  //SDL supports gamepads
+  return true;
+}
+
+int gamepad_get_device_count() {
+  int count = 0;
+
+  for (int i = 0; i < SDL_NumJoysticks(); ++i)
+    if (SDL_IsGameController(i)) count++;
+
+  return count;
+}
+
+bool gamepad_is_connected(int id) { return id >= 0 && id < SDL_NumJoysticks() - 1 && SDL_IsGameController(id); }
```

Comment:
If `SDL_NumJoysticks()` == 1, you are checking `id >= 0` and `id < 0`. Also, are SDL joystick IDs guaranteed to be dense and zero-based? What is the distinction between a Joystick and a GameController?

Reply:
A game controller is a joystick. Sdl has API for both but basically gamepads are mapped to work like Xbox controllers. Where as joystick can have more buttons and axes but aren't consistent

Reply:
fundies is right, but also the GM joystick functions back in like GM5 used DirectInput (before it was deprecated) and the modern ones just use the WinAPI version I believe. The gamepad functions use XInput (on Windows at least) in GMS1.4 and above. I have an Xbox controller to test Windows, but also the gamepad functions in GMS1.4 are supposed to work with a Sixaxis.